src/Validation/URLValidationRESTController.php

```diff
@@ -0,0 +1,278 @@
+	 * @param UserAccess            $dev_tools_user_access UserAccess instance.
+	 */
+	public function __construct( URLValidationProvider $url_validation_provider, UserAccess $dev_tools_user_access ) {
+		$this->namespace               = 'amp/v1';
+		$this->url_validation_provider = $url_validation_provider;
+		$this->dev_tools_user_access   = $dev_tools_user_access;
+	}
+
+	/**
+	 * Registers all routes for the controller.
+	 */
+	public function register() {
+		register_rest_route(
+			$this->namespace,
+			'/validate-post-url/(?P<id>[\d]+)',
+			[
+				'args'   => [
+					'id'            => [
+						'description' => __( 'Unique identifier for the object.', 'amp' ),
+						'required'    => true,
+						'type'        => 'integer',
+					],
+					'preview_id'    => [
+						'description' => __( 'Unique identifier for the preview.', 'amp' ),
+						'required'    => false,
+						'type'        => 'integer',
+					],
+					'preview_nonce' => [
+						'description' => __( 'Preview nonce string.', 'amp' ),
+						'required'    => false,
+						'type'        => 'string',
+					],
+				],
+				[
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::CREATABLE ),
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => [ $this, 'validate_post_url' ],
+					'permission_callback' => [ $this, 'create_item_permissions_check' ],
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+
+		// @todo Additional endpoint to validate a URL (from a URL rather than a post ID).
+	}
+
+	/**
+	 * Checks whether the current user can view AMP validation results.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has permission; WP_Error object otherwise.
+	 */
+	public function create_item_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! $this->dev_tools_user_access->is_user_enabled() ) {
+			return new WP_Error(
+				'amp_rest_no_dev_tools',
+				__( 'Sorry, you do not have access to dev tools for the AMP plugin for WordPress.', 'amp' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns validation information about a URL, validating the URL along the way.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function validate_post_url( $request ) {
+		$post_id       = (int) $request['id'];
+		$preview_id    = (int) $request['preview_id'];
+		$preview_nonce = $request['preview_nonce'];
+		$url           = amp_get_permalink( $post_id );
+
+		if ( ! empty( $preview_id ) && ! empty( $preview_nonce ) ) {
+			$url = add_query_arg(
+				[
+					'preview'       => 1,
+					'preview_id'    => $preview_id,
```

Comment:
Per above, I think this might be simplified to:

```suggestion
		if ( ! empty( $preview_nonce ) ) {
			$url = add_query_arg(
				[
					'preview'       => 1,
					'preview_id'    => $post_id,
```

Reply:
Addressed (along with a general cleanup after `preview_id` removal) in 6fe44f9.